pkg/storage/mvcc_test.go

```diff
@@ -5884,6 +5889,209 @@ func TestMVCCGarbageCollectRangesFailures(t *testing.T) {
 	}
 }
 
+func TestMVCCGarbageCollectClearRange(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+	ctx := context.Background()
+
+	mkKey := func(k string) roachpb.Key {
+		return append(keys.SystemSQLCodec.TablePrefix(42), k...)
+	}
+	rangeStart := mkKey("")
+	rangeEnd := rangeStart.PrefixEnd()
+
+	// Note we use keys of different lengths so that stats accounting errors
+	// would not obviously cancel out if right and left bounds are used
+	// incorrectly.
+	keyA := mkKey("a")
+	keyB := mkKey("bb")
+	keyC := mkKey("ccc")
+	keyD := mkKey("dddd")
+
+	mkTs := func(wallTimeSec int64) hlc.Timestamp {
+		return hlc.Timestamp{WallTime: time.Second.Nanoseconds() * wallTimeSec}
+	}
+
+	ts2 := mkTs(2)
+	ts3 := mkTs(3)
+	ts4 := mkTs(4)
+	tsGC := mkTs(5)
+	tsMax := mkTs(9)
+
+	mkGCReq := func(start roachpb.Key, end roachpb.Key) roachpb.GCRequest_GCClearRangeKey {
+		return roachpb.GCRequest_GCClearRangeKey{
+			StartKey: start,
+			EndKey:   end,
+		}
+	}
+
+	testData := []struct {
+		name string
+		// Note that range test data should be in ascending order (valid writes).
+		before  rangeTestData
+		request roachpb.GCRequest_GCClearRangeKey
+		// Optional start and end range for tests that want to restrict default
+		// key range.
+		rangeStart roachpb.Key
+		rangeEnd   roachpb.Key
+	}{
+		{
+			name: "subset range tombstone",
+			before: rangeTestData{
+				pt(keyB, ts2),
+				rng(keyB, keyC, ts3),
+			},
+			request: mkGCReq(keyA, keyD),
+		},
+		{
+			name: "full range tombstone",
+			before: rangeTestData{
+				pt(keyB, ts2),
+				rng(keyA, keyD, ts4),
+			},
+			request: mkGCReq(keyA, keyD),
+		},
+	}
+	for _, engineImpl := range mvccEngineImpls {
+		t.Run(engineImpl.name, func(t *testing.T) {
+			for _, d := range testData {
+				t.Run(d.name, func(t *testing.T) {
+					engine := engineImpl.create()
+					defer engine.Close()
+
+					// Populate range descriptor defaults.
+					if len(d.rangeStart) == 0 {
+						d.rangeStart = rangeStart
+					}
+					if len(d.rangeEnd) == 0 {
+						d.rangeEnd = rangeEnd
+					}
+
+					var ms enginepb.MVCCStats
+					d.before.populateEngine(t, engine, &ms)
+
+					require.NoError(t,
+						MVCCGarbageCollectWholeRange(ctx, engine, &ms, d.request.StartKey, d.request.EndKey, tsGC),
+						"failed to run mvcc range tombstone garbage collect")
+
+					rks := scanRangeKeys(t, engine)
+					require.Empty(t, rks)
+					ks := scanPointKeys(t, engine)
+					require.Empty(t, ks)
+
+					ms.AgeTo(tsMax.WallTime)
+					it := engine.NewMVCCIterator(MVCCKeyAndIntentsIterKind, IterOptions{
+						KeyTypes:   IterKeyTypePointsAndRanges,
+						LowerBound: d.rangeStart,
+						UpperBound: d.rangeEnd,
+					})
+					expMs, err := ComputeStatsForIter(it, tsMax.WallTime)
+					require.NoError(t, err, "failed to compute stats for range")
+					require.EqualValues(t, expMs, ms, "computed range stats vs gc'd")
+				})
+			}
+		})
+	}
+}
+
+func TestMVCCGarbageCollectClearRangeFailure(t *testing.T) {
```

Comment:
I'd like some tests of inline values too, in particular that we won't accidentally GC them even though they're technically below the GC threshold (timestamp 0).

Reply:
Aren't those values covered by intents? They should have 0 timestamp. At least that's how GC processes them on the queue side. I'll add specific test just in case.

Reply:
Sort of, but the access paths are different enough that it's worth covering both. Intents are stored in a separate lock table, and either require use of an interleaving iterator or by targeting the lock table explicitly, while inline values can show up anywhere.